luarules/gadgets/unit_area_timed_damage.lua

```diff
@@ -12,484 +12,229 @@ function gadget:GetInfo()
-        --for i = 1,#aliveExplosions do
-            if explosionStats.endTime >= currentTime then
-                local x = explosionStats.x
-                local z = explosionStats.z
-                local y = explosionStats.y or Spring.GetGroundHeight(x,z)
-                if explosionStats.cegSpawned == false then
-                    Spring.SpawnCEG(explosionStats.ceg, x, y + 8, z, 0, 0, 0)
-                    explosionStats.cegSpawned = true
-                end
-                local damage = explosionStats.damage*0.733
-                local range = explosionStats.range
-                local resistance = explosionStats.resistance
-                local unitsInRange = Spring.GetUnitsInSphere(x, y, z, range)
-                for j = 1,#unitsInRange do
-                    local unitID = unitsInRange[j]
-                    local unitDefID = Spring.GetUnitDefID(unitID)
-                    if (not UnitDefs[unitDefID].canFly) and (not (UnitDefs[unitDefID].customParams and UnitDefs[unitDefID].customParams.areadamageresistance and string.find(UnitDefs[unitDefID].customParams.areadamageresistance, resistance))) then
-                        local health = Spring.GetUnitHealth(unitID)
-                        if health > damage then
-                            Spring.SetUnitHealth(unitID, health - damage)
-                        else
-                            Spring.DestroyUnit(unitID, false, false)
-                        end
-                        local ux, uy, uz = Spring.GetUnitPosition(unitID)
-                        Spring.SpawnCEG(explosionStats.damageCeg, ux, uy + 8, uz, 0, 0, 0)
-                    end
-                end
-                local featuresInRange = Spring.GetFeaturesInSphere(x, y, z, range)
-                for j = 1,#featuresInRange do
-                    local featureID = featuresInRange[j]
-                    local health = Spring.GetFeatureHealth(featureID)
-                    if health > damage then
-                        Spring.SetFeatureHealth(featureID, health - damage)
-                    else
-                        Spring.DestroyFeature(featureID)
-                    end
-                    local ux, uy, uz = Spring.GetFeaturePosition(featureID)
-                    Spring.SpawnCEG(explosionStats.damageCeg, ux, uy + 8, uz, 0, 0, 0)
-                end
-            else -- This explosion is outdated, we can remove it from the list
-                aliveExplosions[explosionID] = nil
-            end
-        end
-    end
+	UpdateTimedAreas(frame)
+end
+
+function gadget:UnitPreDamaged(unitID, unitDefID, unitTeam, damage, paralyzer, weaponDefID, projID, attackID, attackDefID, attackTeam)
+	local params = TimedAreaDefs[weaponDefID]
+	if params then
+		local unitDef = UnitDefs[unitDefID]
+		local immunity = unitDef.customParams.areadamageresistance
+		if immunity and string.find(immunity, params.atd_resistance, 1, true) then
```

Comment:
```suggestion
		if immunity and immunity == params.atd_resistance then
```
Is this not just an exact string match? Immunity sounds like a good candidate for a bitmask.

Reply:
Saved best for last, done here.